### DISCUSSION TITLE
Get access-token received from OAuth2 provider

### DISCUSSION THREAD

**Discussion by spy16** [pocketbase/pocketbase#654] [Q&A] [answered]
Is there a way to get the access token that was received from GitHub (or any other OAuth2 provider in general) on server-side? 

For context: I am using Pocketbase as library. My app needs to make API calls to GitHub. So I need to have the access token that was issued.

**Comment by ganigeorgiev** [comment, marked as answer]
No. The OAuh2 token is not stored anywhere. The OAuth2 authentication is just a "intermediate" state to associate/create your PocketBase user and nothing more.

Also note that even if you had access to the issued token, the token scopes may not be sufficient for your use case (for github we are using only `"read:user", "user:email"`).

You have to issue another code/token with the correct scopes that are required by your application. You can use directly the std go [`oauth2`](https://pkg.go.dev/golang.org/x/oauth2) package to instantiate your client (the client id and secret could be accessed from `app.Settings().GithubAuth.*`).

---

_**Update:** We now return the raw user provider response and the access token as part of the `meta` props. For more details see https://github.com/pocketbase/pocketbase/discussions/654#discussioncomment-4273677._

**Reply by spy16** [reply]
Okay, understood. Thanks for the quick response.

**Comment by jimafisk** [comment]
I was hoping to get an oauth access token as well. My project uses [Plenti](https://github.com/plentico/plenti) as an SSG for the website content and PocketBase as the database backend for members. Plenti ships with a git-backed CMS that writes directly from the website to a GitLab repo to update content. It uses PKCE to authenticate to GitLab, [similar to PocketBase](https://github.com/pocketbase/pocketbase/issues/55).

I'm trying to unify the login so you don't have to authenticate to GitLab separately for both Plenti and PocketBase. I'm using a single GitLab application (OAuth provider) with both "api" (for Plenti) and "read_user" (for PocketBase) scopes and each login works independently. Trying to create the Plenti user that can write back to the repo from the PocketBase workflow doesn't seem like it's possible without an access token. Is there a way to create / authenticate a PocketBase user from a custom OAuth implementation? Is there a way to perform something like `$client.collection('my_user_type').authWithOAuth2()` using provider info that didn't originate in PocketBase?

**Reply by ganigeorgiev** [reply]
@jimafisk 

> Is there a way to create / authenticate a PocketBase user from a custom OAuth implementation? Is there a way to perform something like $client.collection('my_user_type').authWithOAuth2() using provider info that didn't originate in PocketBase?

No, unfortunately this is not possible at the moment.
I'm not really sure what would be the best approach in this scenario.

The first option would be to get the access token from Plenti and then authenticate the user by sending it to a custom PocketBase endpoint, eg. something like. `$client.collection('my_user_type').authWithOAuth2Token("gitlab", "TOKEN")`.
But different applications have different scope requirements and the above may not always work causing eventual confusions down the road.

This is true also for the opposite - returning the OAuth2 token after PocketBase authentication under the `meta` prop alongside the other OAuth2 profile data. ~~This unfortunately has one additional drawback from security standpoint on mobile devices because not all providers has PKCE support (eg. Twitch) and depending where the redirect is resolved it may not be a good idea to return the 3rd party token in the response.~~ _(this actually may not be such an issue because if the provider doesn't have a PKCE then a malicious user can intercept the redirect url and exchange the code for token on their own; for the next release I'll consider returning the OAuth2 token in the auth success response under the `meta` key)_. 

---

One hacky workaround could be to make use of the fact that most OAuth2 providers uses session cookies and once the user authenticates it will remain login for some time and it will not require a consequent consent from the user (if the scopes for both applications are the same).

In this case, the user can click on the Plenti's "Login with GitLab" button -> complete the authentication and then for PocketBase in the background you can open the signin url in a new small browser window and auto close it on success and continue the usual flow, eg.:

```js
window.open(
    "provider_signin_url", // eg. the (authUrl + redirectUrl) from the providers listAuthMethods() list
    "my_new_window",
    'width=1000,height=500,resizable,menubar=no'
);
```

The redirect handling will be performed in the opened window (you can show a "Loading" spinner for example) and on `authWithOAuth2()` success close the window, eg:

```js
window.opener.location.reload();
window.opener.focus();
window.close();
```

**Reply by jimafisk** [reply]
> for the next release I'll consider returning the OAuth2 token in the auth success response under the meta key

That would be awesome thanks :). Would trying to change the scope clientside work?

```js
let redirectUrl = 'https://localhost:3000/';
const staffLogin = async () => {
  try {
    const authMethods = await $client.collection('staff').listAuthMethods();
    for (const provider of authMethods.authProviders) {
      provider.scope = "api";
      localStorage.setItem('provider', JSON.stringify(provider));
      window.location.href = provider.authUrl + redirectUrl;
    }
  } catch (e) {
    console.log(e);
  }
}
```

**Reply by ganigeorgiev** [reply]
>  Would trying to change the scope clientside work?

The scopes are part of the generated authUrl and you should be able to change them by updating the query params, but I don't remember whether there was a requirement to pass the scopes too when exchanging the code for token (aka. in the redirect handler). I've added it my todo to check it and will consider it while adding the OAuth2 token to the auth response since this was also requested several times already.

**Reply by ganigeorgiev** [reply]
I've just checked the spec - no, they don't need to be sent with the exchange call (the exchange call requires `grant_type`, `code`, `redirect_uri` and eventually the `code_verifier`).

Or in other words, you should be able to pass custom scopes to the `provider.authUrl` by doing something like:
```js
const url = new URL(provider.AuthUrl + redirectUrl);
const params = url.searchParams;
params.set("scope", "myCustomScope")
url.search = params.toString();
const authUrl = url.toString();
```

**Reply by jimafisk** [reply]
Ok great! I did some testing and it seems to work on my end updating the scope clientside before saving the provider info to local storage. By "work" I mean it logs me into the PB backend, not sure if you'll have to do additional work to make sure the access token is scoped correctly. Thanks @ganigeorgiev!

**Edit:** Just seeing your response, yes that makes sense that I'd need to update the `provider.authUrl`. Thanks for the code snippet!

**Reply by ganigeorgiev** [reply]
@jimafisk, @spy16 I've updated the `meta` field response and now it includes 2 new properties:
- `rawUser` - the raw user data response
- `accessToken` - the OAuth2 access token

The final response will look something like this:
```js
{
  "token": "...",
  "record": {
    "collectionId": "_pb_users_auth_",
    "collectionName": "users",
    "created": "2022-11-30 11:49:49.073Z",
    "updated": "2022-11-30 11:49:49.073Z",
    "email": "test@example.com",
    "emailVisibility": false,
    "id": "v5a3muhiwtt5ch9",
    "username": "test123",
    "verified": true,
    "expand": {}
  },
  "meta": {
    "id": "123456789",
    "name": "John Doe",
    "username": "test123",
    "email": "test@example.com",
    "avatarUrl": "",
    "accessToken": "gho_test_token",
    "rawUser": {
      "id": 1234567,
      "collaborators": 0,
      "company": null,
      ...
    }
  }
}
```
Please note that the fields inside `rawUser` are provider specific and the response could change depending on the sent scopes.

The changes are currently only in `master` and they will be available with the next release.
I'm still not sure about the `rawUser` property name so if anyone think that it is not clear and/or has other suggestions please let me know.

**Reply by jimafisk** [reply]
Hi @ganigeorgiev,

I'm using the new [v0.9.0 release](https://github.com/pocketbase/pocketbase/releases/tag/v0.9.0), but I'm not seeing `accessToken` or `rawUser` under the `meta` key (I do see `avatarUrl`, `email`, `id`, `name` and `username`). I'm getting this from the `authData` returned by:

```js
authData = await $client.collection('my_user_type').authWithOAuth2(
  provider.name,
  params.get('code'),
  provider.codeVerifier,
  redirectUrl
);
```

Should I be doing something else to gather this info? Thank you!

**Reply by ganigeorgiev** [reply]
@jimafisk No, you don't need to do anything else. All providers should return those fields (even if they are empty).

**Reply by ganigeorgiev** [reply]
@jimafisk I've just tested with a GitLab OAuth2 app and the `rawUser` and `accessToken` fields are included in the response (I'm using the public GitLab service, aka. https://gitlab.com/).

**Reply by jimafisk** [reply]
@ganigeorgiev - this is indeed working, terribly sorry for wasting your time. I was testing on a deployed environment and needed to `systemctl restart pocketbase` to restart the service after pulling in the changes. Thanks for taking a look!

**Comment by jimafisk** [comment]
Hi @ganigeorgiev, the OAuth2 `accessToken` has been working great, but expires rather quickly. Is there any way to get a refresh token? Thanks!

**Reply by ganigeorgiev** [reply]
The `refreshToken` should be available together with `accessToken` under the `meta` prop.

I've just tested it with GitLab and it seems correct, but I guess it could be empty for some providers if they don't support it or require some special mechanism to issue it.
